src/core/opamSystem.ml

```diff
@@ -1633,6 +1633,40 @@ let translate_patch ~dir orig corrected =
   end;
   close_in ch
 
+let gpatch = lazy begin
+  let rec search_gpatch = function
+    | [] -> None
+    | patch_cmd::patch_cmds ->
+      match OpamProcess.run (make_command ~name:"patch" patch_cmd ["--version"]) with
+      | r ->
+        (match OpamProcess.is_success r, r.OpamProcess.r_stdout with
+         | true, full::_ when
+             OpamStd.String.is_prefix_of ~from:0 ~full "GNU patch " ->
+           Some patch_cmd
+         | _ ->
+           search_gpatch patch_cmds)
+      | exception _ -> search_gpatch patch_cmds
+  in
+  let default_cmd, other_cmds =
+    match OpamStd.Sys.os () with
+    | DragonFly
+    | FreeBSD
+    | NetBSD
+    | OpenBSD -> ("gpatch", ["patch"])
+    | Cygwin
+    | Darwin
+    | Linux
+    | Unix
+    | Win32
+    | Other _ -> ("patch", ["gpatch"])
```

Comment:
the problem with that is that it changes the behaviour compared to 2.1.5

Reply:
yes, well, or even:
```OCaml
let default_cmd, other_cmds = "gpatch", [ "patch" ]
```

avoiding the match on the os type at all. But I hear your "changes the behaviour". With that in mind, I like your PR very much.

The only suggestion left is:
```OCaml
  let default_cmd, other_cmds =
    match OpamStd.Sys.os () with
    | DragonFly
    | FreeBSD
    | NetBSD
    | OpenBSD
    | Darwin -> ("gpatch", ["patch"])
    | Cygwin
    | Linux
    | Unix
    | Win32
    | Other _ -> ("patch", [])
```

so, put Darwin into the pot where gpatch is tried first, and don't try gpatch on Linux & Windows.

But as said, maybe the current state is best for avoiding too much trouble / regressions.

Reply:
Actually considering the code and platforms, maybe

```OCaml
let default_cmd, other_cmds = "patch", [ "gpatch" ]
```

is suitable? This would on OpenBSD and FreeBSD call `patch --version`, and then go to using `gpatch`...